File: menetrend.cpp
#include <iostream>
#include <fstream>
#include <sstream>

using namespace std;
struct Tido {
    int ora, perc, ido;
};
struct Tallomas {
    Tido ind, erk;
};
int main()
{
    // 1. feladat
    Tallomas sz[21][21];
    ifstream be("vonat.txt");
    int szerelveny, allomas, ora, perc;
    string muvelet;
    int i=0;
    while(be >> szerelveny >> allomas >> ora >> perc >> muvelet)
    {
        if(muvelet=="E")
        {
            sz[szerelveny][allomas].erk.ora=ora;
            sz[szerelveny][allomas].erk.perc=perc;
            sz[szerelveny][allomas].erk.ido=ora*60+perc;
        }
        else
        {
            sz[szerelveny][allomas].ind.ora=ora;
            sz[szerelveny][allomas].ind.perc=perc;
            sz[szerelveny][allomas].ind.ido=ora*60+perc;
        }
    }
    // 2. feladat
    cout << "2. feladat" << endl;
    cout << "Az allomasok szama: " << allomas+1 << endl;
    cout << "A vonatok szama: " << szerelveny << endl;

    // 3. feladat
    cout << "3. feladat" << endl;
    int maxvarakozas=-1;
    int maxszerelveny, maxallomas;
    for(int i=1; i<=szerelveny; i++)
    {
        for(int j=1; j<allomas; j++)
        {
            int varakozas=sz[i][j].ind.ido-sz[i][j].erk.ido;
            if(varakozas>maxvarakozas)
            {
                maxszerelveny=i;
                maxallomas=j;
                maxvarakozas=varakozas;
            }
        }
    }
    cout << "A(z) " << maxszerelveny << ". vonat a(z) " << maxallomas << ". allomason " << maxvarakozas << " percet allt." << endl;

    // 4. feladat
    cout << "4. feladat" << endl;
    int vsz, vora, vperc;
    cout << "Adja meg egy vonat azonositojat! ";
    cin >> vsz;
    cout << "Adjon meg egy idopontot (ora perc)! ";
    cin >> vora >> vperc;

    // 5. feladat
    cout << "5. feladat" << endl;
    int hossz=sz[vsz][allomas].erk.ido-sz[vsz][0].ind.ido;
    cout << "A(z) "<< vsz << ". vonat utja ";
    if(hossz>2*60+22) cout << hossz-(2*60+22) << " perccel hosszabb volt az eloirtnal.";
    if(hossz==2*60+22) cout << " pontosan az eloirt ideig tartott.";
    if(hossz<2*60+22) cout << (2*60+22)-hossz << " perccel rovidebb volt az eloirtnal.";
    cout << endl;

    // 6. feladat
    stringstream konvert;
    konvert << "halad" << vsz << ".txt";
    string fajlnev;
    konvert >> fajlnev;
    ofstream ki(fajlnev);
    for(int i=1; i<=allomas; i++)
    {
        ki << i << ". allomas: " << sz[vsz][i].erk.ora << ":" << sz[vsz][i].erk.perc << endl;
    }
    ki.close();

    // 7. feladat
    cout << "7. feladat" << endl;
    int idopont=vora*60+vperc;
    for(int i=1; i<=szerelveny; i++)
    {
        if(sz[i][0].ind.ido<=idopont and idopont<sz[i][1].erk.ido)
        {
            cout << "Az " << i << ". vonat a " << 1 << ". es a " << 2 << ". allomas kozott jart." << endl;
        }
        for(int j=1; j<allomas; j++)
        {
            if(sz[i][j].erk.ido<=idopont and idopont<sz[i][j].ind.ido)
            {
                cout << "Az " << i << ". vonat a " << j << ". allomason allt." << endl;
            }
            if(sz[i][j].ind.ido<=idopont and idopont<sz[i][j+1].erk.ido)
            {
                cout << "Az " << i << ". vonat a " << j << ". es a " << j+1 << ". allomas kozott jart." << endl;
            }
        }
    }
    return 0;
}

File: menetrend.py
import datetime


class Vonat:
    def __init__(self, sor):
        m = sor.split('\t')
        self.azon = int(m[0])
        self.allomas = int(m[1])
        self.ido = datetime.datetime(2020, 1, 1, int(m[2]), int(m[3]))
        self.status = str(m[4])
        self.varakozas = -1


def eltelt_ido_min(v1:Vonat, v2:Vonat):
    return int((v2.ido - v1.ido).total_seconds()) // 60


def main():
    # 1. feladat
    sr = open('vonat.txt', 'r', encoding='UTF8')
    vonatok = list()
    for i in sr.read().splitlines():
        akt_vonat = Vonat(i)
        # várakozási idő [min] meghatározása indulásoknál:
        if akt_vonat.status == 'I' and akt_vonat.allomas != 0: # Ha nem az első állomásról indul
            elozo_adatok = list(filter(lambda x: x.azon == akt_vonat.azon, vonatok))
            akt_vonat.varakozas =  eltelt_ido_min(elozo_adatok[-1], akt_vonat)
        vonatok.append(akt_vonat)
    sr.close()

    # 2. feladat
    vonatok_h = set()
    allomasok_h = set()
    for i in vonatok:
        vonatok_h.add(i.azon)
        allomasok_h.add(i.allomas)
    print(f'2. feladat\nAz állomások száma: {len(allomasok_h)}\nA vonatok száma: {len(vonatok_h)}')

    # 3. feladat
    rendezve_varakozas = sorted(vonatok, key=lambda x: x.varakozas)
    maxv = rendezve_varakozas[-1]
    print(f'3. feladat\nA(z) {maxv.azon}. vonat a(z) {maxv.allomas}. állomáson {maxv.varakozas} percet állt.')

    input_vonat = int(input('4. feladat\nAdja meg egy vonat azonosítóját! '))
    input_ido = input('Adjon meg egy időpontot (óra perc)! ')

    print('5. feladat')
    input_vonat_adatok = list(filter(lambda x: x.azon == input_vonat, vonatok))
    menetido = eltelt_ido_min(input_vonat_adatok[0], input_vonat_adatok[-1])
    if menetido == 142: # 2 * 60 + 22 = 142
        print(f'A(z) {input_vonat}. vonat útja pontosan az előírt ideig tartott.')
    elif menetido > 142:
        print(f'A(z) {input_vonat}. vonat útja {menetido - 142} perccel hosszabb volt az előírtnál.')
    else:
        print(f'A(z) {input_vonat}. vonat útja {142 - menetido} perccel rövidebb volt az előírtnál.')

    # 6. feladat
    sw = open(f'halad{input_vonat}.txt', 'w', encoding='UTF8')
    for i in input_vonat_adatok:
        if i.status == 'E':
            sw.write(f'{i.allomas}. állomás: {i.ido.hour}:{i.ido.minute}\n')
    sw.close()

    print('7. feladat')
    ido = datetime.datetime(2020, 1, 1, int(input_ido.split(' ')[0]), int(input_ido.split(' ')[1]))
    for i in vonatok_h:
        av = list(filter(lambda x: x.azon == i, vonatok)) # aktuális vonat
        for j in range(1, len(av)):
            if av[j - 1].status == 'I': # j - 1 indulási idő, j a következő érkezés ideje
                if av[j - 1].ido <= ido < av[j].ido:
                    print(f'A(z) {av[j].azon}. vonat a {av[j-1].allomas}. és a {av[j].allomas}. állomás között járt.')
            else:
                if av[j - 1].ido <= ido < av[j].ido: # j - 1 érkezési idő, j a következő indulás ideje
                    print(f'A(z) {av[j].azon}. vonat a {av[j].allomas}. állomáson állt.')


if __name__ == '__main__':
    main()
